accession-commons-core/src/main/java/uk/ac/ebi/ampt2d/commons/accession/core/DecoratedAccessioningService.java

```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Override
+    public List<AccessionWrapper<MODEL, HASH, ACCESSION>> get(List<? extends MODEL> accessionedObjects) {
+        return decorate(service.get(accessionedObjects));
+    }
+
+    @Override
+    public List<AccessionWrapper<MODEL, HASH, ACCESSION>> getByAccessions(List<ACCESSION> accessions) {
+        return decorate(service.getByAccessions(parse(accessions)));
+    }
+
+    private List<DB_ACCESSION> parse(List<ACCESSION> accessions) {
+        return accessions.stream().map(parsingFunction).collect(Collectors.toList());
+    }
+
+    @Override
+    public AccessionWrapper<MODEL, HASH, ACCESSION> getByAccessionAndVersion(ACCESSION accession, int version)
+            throws AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        return decorate(service.getByAccessionAndVersion(parse(accession), version));
+    }
+
+    private DB_ACCESSION parse(ACCESSION accession) {
+        return parsingFunction.apply(accession);
+    }
+
+    @Override
+    public AccessionVersionsWrapper<MODEL, HASH, ACCESSION> update(ACCESSION accession, int version, MODEL message)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        return decorate(service.update(parse(accession), version, message));
+    }
+
+    private AccessionVersionsWrapper<MODEL, HASH, ACCESSION> decorate(
+            AccessionVersionsWrapper<MODEL, HASH, DB_ACCESSION> accessionVersions) {
+        return new AccessionVersionsWrapper<>(decorate(accessionVersions.getModelWrappers()));
+    }
+
+    @Override
+    public AccessionVersionsWrapper<MODEL, HASH, ACCESSION> patch(ACCESSION accession, MODEL message)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        return decorate(service.patch(parse(accession), message));
+    }
+
+    @Override
+    public void deprecate(ACCESSION accession, String reason) throws AccessionMergedException,
+            AccessionDoesNotExistException, AccessionDeprecatedException {
+        service.deprecate(parse(accession), reason);
+    }
+
+    @Override
+    public void merge(ACCESSION accessionOrigin, ACCESSION mergeInto, String reason) throws AccessionMergedException,
+            AccessionDoesNotExistException, AccessionDeprecatedException {
+        service.merge(parse(accessionOrigin), parse(mergeInto), reason);
+    }
+
+    public static <MODEL, HASH, DB_ACCESSION> DecoratedAccessioningService<MODEL, HASH, DB_ACCESSION, String>
+    buildPrefixAccessionService(AccessioningService<MODEL, HASH, DB_ACCESSION> service, String prefix,
+                                Function<String, DB_ACCESSION> parseFunction) {
+        return new DecoratedAccessioningService<>(service, accession -> prefix + accession,
+                s -> parseFunction.apply(s.substring(prefix.length())));
```

Comment:
I think you should check the prefix of the decorated accession matches the `prefix` parameter here before calling `substring`. Otherwise you can get the accession with different prefix decorated accessions.

Reply:
That is an excelent point!